clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -275,8 +275,10 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
                 for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses().entrySet()) {
                     TopicPartition tp = entry.getKey();
                     ProduceResponse.PartitionResponse partResp = entry.getValue();
-                    ProducerBatch batch = batches.get(tp);
-                    completeBatch(batch, partResp, correlationId, now);
+                    if (tp != null && partResp != null) {
+                        ProducerBatch batch = batches.get(tp);
+                        completeBatch(batch, partResp, correlationId, now);
+                    }
```

Comment:
Have you verified that `tp` and `partResp` can be `null`? And, if so, what was the broker and client jar versions? This would imply a bug in the broker.